### PR TITLE
tests: Build images lazily

### PR DESCRIPTION
**Pull request by Jongy**
Closes: #480

### CODE REVIEW

File: tests/conftest.py [hunk 1]
```diff
@@ -233,7 +234,12 @@ def application_docker_images(docker_client: DockerClient) -> Mapping[str, Image
         for tag, args in tags_listing.items():
             name = image_name(runtime, tag)
             assert name not in images
-            images[name] = _build_image(docker_client, runtime, **args)
+
+            @lru_cache(maxsize=1)
+            def image_builder() -> Image:
+                return _build_image(docker_client, runtime, **args)
```
Comment (Jongy) [inline tests/conftest.py:240]: Eh this won't work because `runtime` and `args` get re--bound

**Review by netaneld122 — DISMISSED** [review]

File: tests/conftest.py [hunk 2]
```diff
@@ -183,6 +183,7 @@ def gprofiler_docker_image(docker_client: DockerClient) -> Iterable[Image]:
     yield docker_client.images.get("gprofiler")
 
 
+@lru_cache(maxsize=1024)
```
Comment (Jongy) [inline tests/conftest.py:186]: I implemented a cache this way, another option would be to use `fixture(scope="session") on `application_docker_image` but I prefer to have an explicit cache

**Review by netaneld122 — DISMISSED** [review]

**Comment by Jongy** [comment]
Bleh dicts are not cachable, I could use https://stackoverflow.com/a/53394430/797390 but I'll think of something else

**Comment by Jongy** [comment]
Meh, I'll use `scope="session"`.

**Review by netaneld122 — DISMISSED** [review]

**Comment by Jongy** [comment]
```
ScopeMismatch: You tried to access the 'function' scoped fixture 'runtime' with a 'session' scoped request object, involved factories
```

kill me

**Comment by Jongy** [comment]
```
In [26]: %timeit -n 100 docker_client.images.build(path=".", rm=True, dockerfile="Dockerfile")[0]                                                                                                          
40.1 ms ± 1.26 ms per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

This is from building `tests/containers/java`. So Docker does enough caching for us.

**Comment by Jongy** [comment]
This is same as 
image

**Review by netaneld122 — APPROVED** [review]

**Comment by Jongy** [comment]
This already passed as https://github.com/Granulate/gprofiler/pull/486/commits/5a26241562ea73586bb7ed1a962d1f7277869387

**Pull request merged**